### PR TITLE
Auto-convert binary arrays to strings in read_nexus

### LINKED ISSUE #420: [Bug]: read_nexus cannot process files with fixed-length strings
### Contact Details

rettig@fhi-berlin.mpg.de

### What happened?

When reading a file with read_nexus, that contains fixed-length strings, these are returned as byte arrays, and the program terminates.

Relevant entries in the file look like:
grafik

### Relevant log output

```shell
rettig@pcr840:..AreaB/transformations_testing> read_nexus -f NXmpes_liquid\ test\ -\ 2024-07-16.nxs 
Traceback (most recent call last):
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/.pyenv/bin/read_nexus", line 8, in <module>
    sys.exit(main())
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/.pyenv/lib/python3.8/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/.pyenv/lib/python3.8/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/.pyenv/lib/python3.8/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/.pyenv/lib/python3.8/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/pynxtools/src/pynxtools/nexus/nexus.py", line 843, in main
    nexus_helper.process_nexus_master_file(None)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/pynxtools/src/pynxtools/nexus/nexus.py", line 777, in process_nexus_master_file
    self.full_visit(self.in_file, self.in_file, "", self.visit_node)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/pynxtools/src/pynxtools/nexus/nexus.py", line 761, in full_visit
    self.full_visit(root, child, full_name, func)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/pynxtools/src/pynxtools/nexus/nexus.py", line 756, in full_visit
    func(name, hdf_node)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/pynxtools/src/pynxtools/nexus/nexus.py", line 700, in visit_node
    process_node(hdf_node, "/" + hdf_name, self.parser, self.logger)
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/pynxtools/src/pynxtools/nexus/nexus.py", line 384, in process_node
    f"===== GROUP (/{hdf_path} "
  File "/mnt/pcshare/users/Laurenz/AreaB/transformations_testing/pynxtools/src/pynxtools/nexus/nexus.py", line 59, in get_nx_class_path
    get_nx_class_path(get_hdf_info_parent(hdf_info))
TypeError: can only concatenate str (not "numpy.bytes_") to str
```

### PR DESCRIPTION
**Pull request by rettigl**
add try-except blocks to auto decode byte strings (from fixed-length strings)

Closes #420

### CODE REVIEW

**Comment by rettigl** [comment]
Not sure if all these changes make sense or there might be a better option. I can provide a test file if you need.

File: src/pynxtools/nexus/nexus.py [hunk 1]
```diff
@@ -55,15 +65,26 @@ def get_nx_class_path(hdf_info):
...
+            )
+        except AttributeError:
+            return (
+                get_nx_class_path(get_hdf_info_parent(hdf_info))
+                + "/"
+                + (
+                    hdf_node.attrs["NX_class"]
+                    if "NX_class" in hdf_node.attrs.keys()
+                    else hdf_node.name.split("/")[-1]
+                )
```
Comment (lukaspie) [inline src/pynxtools/nexus/nexus.py:86]: Maybe possible to shorten like this?
```suggestion
        nx_class = hdf_node.attrs.get("NX_class", hdf_node.name.split("/")[-1])
        if isinstance(nx_class, bytes):
            nx_class = nx_class.decode("UTF-8")
        return get_nx_class_path(get_hdf_info_parent(hdf_info)) + "/" + nx_class
```

**Review by lukaspie — APPROVED** [review]
LGTM, but I haven't tested it (updated tests seem fine though).

I made some suggestions to avoid the try-except blocks.

File: src/pynxtools/nexus/nexus.py [hunk 2]
```diff
@@ -540,12 +593,20 @@ def signal_helper(nxdata):
...
+                    and sig.attrs.get("signal").decode("UTF-8") == "1"
+                ):
+                    return sig
+            except AttributeError:
+                if (
+                    sig.attrs.get("signal")
+                    and sig.attrs.get("signal") is str
+                    and sig.attrs.get("signal") == "1"
+                ):
+                    return sig
```
[the remark this answers is not in the document]
Comment (rettigl) [reply src/pynxtools/nexus/nexus.py:609]: Thanks, these are all good suggestions. Based on this, I write a small helper function now and inserted this everywhere, which gives even less changes to the code.

File: src/pynxtools/definitions
Comment (rettigl) [inline src/pynxtools/definitions:1]: reminder to change after merging PR in nexus_definitions

File: src/pynxtools/nexus/nexus.py [hunk 3]
```diff
@@ -517,12 +562,20 @@ def nxdata_helper(nxentry):
...
+                and nxentry[key].attrs["NX_class"].decode("UTF-8)") == "NXdata"
+            ):
+                lnxdata.append(nxentry[key])
+        except AttributeError:
+            if (
+                isinstance(nxentry[key], h5py.Group)
+                and nxentry[key].attrs.get("NX_class")
+                and nxentry[key].attrs["NX_class"] == "NXdata"
+            ):
+                lnxdata.append(nxentry[key])
```
[the remark this answers is not in the document]
Comment (rettigl) [reply src/pynxtools/nexus/nexus.py]: obsolete

**Comment by rettigl** [comment]
@lukaspie I changed the implementation now completely into a separate function. This should make things cleaner. I put this into nxdl_tools, so this requires the change to the definitions commit. I used the commit in the rebased feature branch, this should still be valid after merging, no? The current fairmat definitions fail tests for some plugins.
@sanbrock As you wrote much of this code, I believe, I would wait with merging for your approval as well.

**Comment by rettigl** [comment]
@RonHildebrandt The pynxtools-ellips plugin seems to have the nexus-definitions commit hard-coded in the test data. How would this best be changed, in order to allow the tests to pass before merging?

**Comment by RonHildebrandt** [comment]
@rettigl This was due change exactly yesterday afternoon. I have bring this on in the standup meeting to see, how this can set up correctly to let the test pass. (right now, I have no idea - have to get help=

**Comment by lukaspie** [comment]
@rettigl we want to update the definitions separately today anyway, see #425. We are first consolidating the nexus_definitions fairmat branch (including the PR @RonHildebrandt mentioned), we can get your PR on the `nxdl_utils` in there as well. Once we have #425 in, then we can rebase this PR here (the tests should work then) and merge it afterwards.

Regarding the nexus-definitions commit hard-coded in the test data, I think this should not be there at all. The reader will probably overwrite them anyway (see [this line](https://github.com/FAIRmat-NFDI/pynxtools/blob/555e38ea80a03cc46facca2535d01a0c2223526e/src/pynxtools/dataconverter/convert.py#L177C17-L177C41)), so we can just remove it from the ELN in the test data.

**Comment by lukaspie** [comment]
@rettigl with the inclusion of #428, would you be fine with closing this one?

**Comment by rettigl** [comment]
> @rettigl with the inclusion of #428, would you be fine with closing this one?

Absolutely

**Pull request closed without merging**